### ISSUE TITLE
Unable to generalize a type variable. It is not unranked.

### ISSUE THREAD

**Issue by pdavidow** [elm-lang/elm-compiler#1592]
https://github.com/pdavidow/btree/tree/uhoh

```
$ elm-make Main.elm --output elm.js

It looks like something went wrong with the type inference algorithm.

Unable to generalize a type variable. It is not unranked.

Please create a minimal example that triggers this problem and report it to
<https://github.com/elm-lang/elm-compiler/issues>
elm-make: thread blocked indefinitely in an MVar operation
```

OS: Windows10 64bit
Browser: Chrome Version 58.0.3029.96 (64-bit)
Elm version: 0.18

**Comment by evancz** [comment]
I am consolidating all the examples in #1613, so it is easier to find them all.

Thank you for the report!

**Comment by evancz** [comment]
I gotta have an http://sscce.org to be able to reproduce things easily.

Since almost all of the #1613 issues are resolved by 4b25980d93ada12236c794f35570e197be82f787, so I am going to assume this one is as well. If the example can be simplified into a single file that I can copy/paste, I am happy to check on it though!
